Prevent Transferring to 0x0 and Contract Address

Muhammad-Altabba opened the pull request:
Prevent transferring to the 0x0 address, and the contract address, according to Token Implementation Best Practices.
Reference: https://consensys.github.io/smart-contract-best-practices/tokens/

maurelian commented:
@Muhammad-Altabba I'd ask you to also implement tests to verify the behaviour you've introduced, but first...

@skmgoldin If we added these protections (ie. prevent send to `0x0` and `address(this)`), I think we should do so in a child contract that inherits from EIP20.sol. What do you think?

Muhammad-Altabba commented:
I will be happy to implement tests to verify the behavior. And if we need to add the proposed check in a child contract I will also be happy to implement (even I prefer it to be inside the same EIP20.sol contract).
Please direct me with your recommendation and I will do the needed.
Thanks,

skmgoldin commented:
I feel it definitely should not be in EIP20.sol. Perhaps a different implementation called something like SafeEIP20.sol or something.

Muhammad-Altabba commented:
Dear @maurelian and @skmgoldin,
I added SafeEIP20.sol and changed EIP20Factory to use SafeEIP20. And I wrote JavaScript Truffle Tests for the new contract: SafeEIP20.
Please let me know if you think that more things are needed to be done or modified.
Thanks,

maurelian commented on test/eip20/safeEip20.js:
```diff
@@ -0,0 +1,239 @@
+  });   
+  
+  
+  //Starting from this line, the following tests are the same tests applied to EIP20.
```
Is there a way that these tests could be exported from the other test file, and run on the safeEIP20 implementation? I'd prefer not to have to copy/paste code changes made from one to the next.

Muhammad-Altabba replied:
Upon your comment, I Encapsulated the EIP compatible functions' tests inside a function and make it called for both EIP20 and SafeEIP20 smart contracts. And the duplicated tests code at safeEip20.js has been deleted.
So now, there is no copy/past code.

Muhammad-Altabba commented:
Hello,
Is there anything, I can help with, to speed up accepting this Pull Request?
Thanks,

Muhammad-Altabba commented:
Hi,
Could I help regarding this pull request?

maurelian commented:
Done. Apologies for the delay Muhammad.

Muhammad-Altabba commented:
No problem,
Many thanks,
